File: ios/Sources/LeoKit/Components/LKTable.swift
//
//  LKTable.swift
//  LeoKit
//
//  Created by Leo Ho on 2026/09/20.
//

import SwiftUI

/// 多欄、可以上下比較的資料
///
/// - Note: 排序、分頁與篩選的邏輯都在呼叫端，這個元件只負責畫
/// - Note: 表格是寬版面的元件。iPhone 上請改用 `List` 搭配 `LKListRow`，一列一筆
/// - Note: 需要選取列、拖曳欄寬或跟著系統排序的資料表，用系統的 `Table` 會比這個合適
/// - Note: 數字欄靠右並用等寬數字，讓位數上下對齊
public struct LKTable: View {

    // MARK: - Properties

    /// 這張表在說什麼，唸給螢幕閱讀器聽
    private let caption: String

    /// 每一欄的定義，順序就是顯示順序
    private let columns: [LKTableColumn]

    /// 每一列的資料
    private let rows: [LKTableRow]

    // MARK: - Init

    /// 建立一張表
    ///
    /// - Parameters:
    ///   - caption: 這張表在說什麼
    ///   - columns: 每一欄的定義，順序就是顯示順序
    ///   - rows: 每一列的資料
    public init(_ caption: String, columns: [LKTableColumn], rows: [LKTableRow]) {
        self.caption = caption
        self.columns = columns
        self.rows = rows
    }

    // MARK: - Body

    /// 骨架：一列表頭，然後每一列資料，列與列之間畫分隔線
    public var body: some View {
        Grid(alignment: .leading, horizontalSpacing: LKSpacing.spacing12, verticalSpacing: 0) {
            header

            ForEach(rows) { row in
                divider

                dataRow(row)
            }
        }
        .padding(.horizontal, LKSpacing.spacing16)
        .background(LKColor.bgSurface)
        .accessibilityLabel(caption)
    }
}

// MARK: - Private Views

private extension LKTable {

    /// 表頭那一列
    var header: some View {
        GridRow {
            ForEach(columns) { column in
                Text(column.label)
                    .frame(maxWidth: .infinity, alignment: column.isNumeric ? .trailing : .leading)
                    .font(LKFont.subhead)
                    .foregroundStyle(LKColor.textSecondary)
                    .padding(.vertical, LKSpacing.spacing8)
            }
        }
    }

    /// 一列資料
    ///
    /// - Parameter row: 要畫的那一列
    /// - Returns: 一列的每一格
    func dataRow(_ row: LKTableRow) -> some View {
        GridRow {
            ForEach(columns) { column in
                cell(row.text(for: column), isNumeric: column.isNumeric)
            }
        }
    }

    /// 一格資料
    ///
    /// - Parameters:
    ///   - text: 這一格的文字
    ///   - isNumeric: 是否為數字欄
    /// - Returns: 一格
    @ViewBuilder
    func cell(_ text: String, isNumeric: Bool) -> some View {
        if isNumeric {
            Text(text)
                .frame(maxWidth: .infinity, alignment: .trailing)
                .padding(.vertical, LKSpacing.spacing12)
                .font(LKFont.callout)
                .foregroundStyle(LKColor.textPrimary)
                .monospacedDigit()
        } else {
            Text(text)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.vertical, LKSpacing.spacing12)
                .font(LKFont.callout)
                .foregroundStyle(LKColor.textPrimary)
        }
    }

    /// 列與列之間的分隔線，橫跨所有欄
    var divider: some View {
        Rectangle()
            .fill(LKColor.borderSubtle)
            .frame(height: LKSize.hairline)
            .gridCellColumns(columns.count)
    }
}

// MARK: - Preview

#Preview("本月訂閱") {
    LKTable(
        "本月訂閱與金額",
        columns: [
            LKTableColumn(key: "name", label: "服務"),
            LKTableColumn(key: "cycle", label: "週期"),
            LKTableColumn(key: "amount", label: "金額", isNumeric: true),
        ],
        rows: [
            LKTableRow(cells: ["name": "Netflix", "cycle": "月繳", "amount": "390"]),
            LKTableRow(cells: ["name": "Spotify", "cycle": "月繳", "amount": "149"]),
            LKTableRow(cells: ["name": "iCloud+", "cycle": "年繳", "amount": "1,090"]),
        ]
    )
    .padding(.vertical, LKSpacing.spacing16)
}

File: android/leokit/src/main/kotlin/io/github/leoho0722/leokit/components/LKTable.kt
package io.github.leoho0722.leokit.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import io.github.leoho0722.leokit.LKTheme
import io.github.leoho0722.leokit.tokens.LKSpacing

/**
 * 多欄、可以上下比較的資料。
 *
 * Material 沒有表格元件，所以這是自繪的 `LazyColumn`。
 * 表格是寬版面的元件：手機上請改用 [LKListRow] 一列一筆，只有平板才擺多欄。
 * 排序、分頁與篩選的邏輯都在呼叫端，這個元件只負責畫。
 * 數字欄靠右並用等寬數字，讓位數上下對齊。
 *
 * @param caption 這張表在說什麼，唸給螢幕閱讀器聽
 * @param columns 每一欄的定義，順序就是顯示順序
 * @param rows 每一列的資料，鍵對應 [LKTableColumn.key]
 * @param modifier 由呼叫端套在整張表上的修飾子
 */
@Composable
public fun LKTable(
    caption: String,
    columns: List<LKTableColumn>,
    rows: List<LKTableRow>,
    modifier: Modifier = Modifier,
) {
    val colors = LKTheme.colors
    val typography = LKTheme.typography

    Column(
        modifier = modifier
            .fillMaxWidth()
            .background(colors.bgSurface)
            .semantics { contentDescription = caption },
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = LKSpacing.spacing16, vertical = LKSpacing.spacing8),
            horizontalArrangement = Arrangement.spacedBy(LKSpacing.spacing12),
        ) {
            columns.forEach { column ->
                Text(
                    text = column.label,
                    style = typography.subhead,
                    color = colors.textSecondary,
                    textAlign = if (column.numeric) TextAlign.End else TextAlign.Start,
                    modifier = Modifier.weight(1f),
                )
            }
        }
        LazyColumn {
            items(rows) { row ->
                HorizontalDivider(color = colors.borderSubtle)
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(
                            horizontal = LKSpacing.spacing16,
                            vertical = LKSpacing.spacing12,
                        ),
                    horizontalArrangement = Arrangement.spacedBy(LKSpacing.spacing12),
                ) {
                    columns.forEach { column ->
                        Text(
                            text = row.cells[column.key].orEmpty(),
                            style = typography.callout,
                            color = colors.textPrimary,
                            textAlign = if (column.numeric) TextAlign.End else TextAlign.Start,
                            modifier = Modifier.weight(1f),
                        )
                    }
                }
            }
        }
    }
}

/** 表格的一欄：它的標題、對應到哪個鍵，以及要不要靠右對齊。 */
public data class LKTableColumn(
    /** 這一欄對應 [LKTableRow.cells] 裡的哪個鍵。 */
    public val key: String,
    /** 欄位標題。 */
    public val label: String,
    /** 是否為數字欄；數字欄靠右，方便上下比較位數。 */
    public val numeric: Boolean = false,
)

/** 表格的一列：每一欄的鍵對到那一格要顯示的文字。 */
public data class LKTableRow(
    /** 鍵對應 [LKTableColumn.key]，值是那一格的文字。 */
    public val cells: Map<String, String>,
)
